Teach kubectl drain to evict on nodes in parallel

mnussbaum opened the pull request:
/kind feature

**What this PR does / why we need it**:

Pod eviction can be a very slow operation. Currently `kubectl drain` takes multiple node arguments and then evicts pods on each given node in serial. This change makes `kubectl drain` time roughly constant instead of linear with the number of nodes given.

**Does this PR introduce a user-facing change?**:

Yes

```release-note
"kubectl drain" with multiple node arguments now drains all nodes in parallel rather then in sequence
```

mnussbaum committed:
Teach kubectl drain to evict on nodes in parallel

Pod eviction can be a very slow operation. Currently kubectl drain
takes multiple node arguments and evicts pods on each node in serial.
This change makes kubectl drain time roughly constant instead of linear
with the number of nodes given.

mnussbaum commented:
/assign @soltysh

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/drain/drain.go:
```diff
@@ -297,34 +298,48 @@ func (o *DrainCmdOptions) RunDrain() error {
-				}
-			}
-			break
+			deleteOrEvictWaitGroup.Done()
```
My gut feeling is this should be `defer deleteOrEvictWaitGroup.Done()` as the first line in the go routine to make sure the wait is always done. You don't return anywhere else, so maybe it is OK.

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/drain/drain.go:
```diff
@@ -297,34 +298,48 @@ func (o *DrainCmdOptions) RunDrain() error {
+
+	deleteOrEvictWaitGroup.Wait()
+	remainingNodes := []string{}
+	deleteOrEvictResultMutex.Lock()
```
Is this mutex necessary? I'm thinking after `deleteOrEvictWaitGroup.Wait()` you are single-threaded now.

mnussbaum replied:
The lock isn't necessary here. It felt odd to read from something protected by a mutex without taking out the mutex's lock, but you're right this isn't necessary at all

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/drain/drain.go:
```diff
@@ -297,34 +298,48 @@ func (o *DrainCmdOptions) RunDrain() error {
 	drainedNodes := sets.NewString()
-	var fatal error
+	var deleteOrEvictErrors []error
+	var deleteOrEvictResultMutex sync.Mutex
```
If you use a channel instead, you won't need `deleteOrEvictErrors`, `deleteOrEvictResultMutex`, or `deleteOrEvictWaitGroup`. Did you consider that way?

mnussbaum replied:
Ah yeah I should've thought to use a channel, that would be nicer. I can revisit the code with that approach

mnussbaum replied:
I'm working through the details of this now and I'm not certain using channels is the best approach. If the goal is to preserve the existing behavior of printing the success or failure of draining each node, and then printing the list of nodes that still need to be drained at the end, and then to return drain errors then I think I need three channels. One for successfully drained nodes, one for unsuccessfully drained nodes and one for errors. I also still need the wait group so that I can block until all nodes have either succeeded or failed at draining.

Is there something I'm missing that could simplify the channel solution?

brianpursley replied:
Hmm, I don't know for sure, but I was originally thinking something like this:
```
var results = make(chan error)
for _, info := range o.nodeInfos {
	info := info
	go func() { results <- o.deleteOrEvictPodsSimple(info) }()
}

for range o.nodeInfos {
	err := <- results
	// TODO: Do everything here that was being done in the original for loop
}
```

But I guess you also need `info` coming out of the channel too, so maybe that wouldn't really make it simpler because you would need a struct or something to return.

Probably fine to use waitgroup in this case

mnussbaum replied:
Gotcha yeah I think the output requirements makes the waitgroup a bit simpler thann the channel option. I can go the struct route though if you'd prefer it

brianpursley replied:
No it’s fine, I was just thinking it might be simpler for a second, but waitgroup is probably fine

soltysh replied:
Either solution would be good, I'd personally lean towards solution proposed by Brian here. You'd need two channels one for errors and another for node names. But that's not something I'd block a PR on.

mnussbaum committed:
Remove unnecessary locking. Defer waitgroup done call

mnussbaum commented:
@brianpursley the test failures all appear unrelated to the change I made. What's the best way forward here?

brianpursley commented:
> @brianpursley the test failures all appear unrelated to the change I made. What's the best way forward here?

The e2e tests are flakey, so you have to retest sometimes.

/retest

mnussbaum commented:
/retest

mnussbaum commented:
/retest

mnussbaum commented:
/retest

mnussbaum commented:
@brianpursley looks like all the tests are finally passing! Thanks for the tip

mnussbaum commented:
@brianpursley is there anything else I can do to help move this forward? Thanks for your help!

brianpursley commented:
/lgtm

I would like a second set of :eyes: on the concurrency but it looks like it should be fine to me.

soltysh reviewed (changes requested):
/hold
The code itself is good, but this changes the behavior of the command. Currently if a drain on a node fails we will break the draining immediately. This approach on the other hand will go through all nodes as far as it can. I'm not sure that's acceptable w/o broader discussion
@michaelgugino @smarterclayton you might have opinions on this, since you looked at this in the past

michaelgugino reviewed (changes requested):
I think the implementation is fine.  If someone wants to run drain serially, they can use a bash loop, though we build in an option to enforce existing behavior, I don't think it's necessary.  I think this is what most users will want.

Since this is a change in behavior, it requires a release note.

My concern is the logging.  On many log lines, the node is omitted.  Since previously we're draining one node at a time, you could determine the context easily.  Now that many drains are happening in parallel, it's going to be hard to decipher which errors belong to which nodes unless we explicitly add node info to log lines.

Speaking of logging, our unit tests are not thread safe, FYI: https://github.com/kubernetes/kubernetes/blob/d420894997f2902fdd1c0bd98629e29a769da9d1/staging/src/k8s.io/kubectl/pkg/cmd/drain/drain_test.go#L215

This doesn't look to be a problem today, but we should leave some comments around the lines using those buffers so they don't get copied somewhere they shouldn't be in the future.

I've looked for other thread-unsafe conditions, but I don't see any.

soltysh commented:
I'd suggest then adding a new flag and user could choose between serial or parallel execution. I think that'd be the best option.

michaelgugino commented:
> I'd suggest then adding a new flag and user could choose between serial or parallel execution. I think that'd be the best option.

Yeah, I think if we make parallel opt-in, then maybe we don't need to worry about the logging being perfect.

mnussbaum commented:
An opt-in flag makes sense, I can take a look at adding that shortly! I'll also see if I can address the logging notes

mnussbaum committed:
Rename wait group var for PR feedback

mnussbaum committed:
Guard parallel drain functionality behind a flag

Users will now need to invoke `kubectl drain --parallelize-nodes ...` to
drain multiple nodes in parallel.

mnussbaum commented on staging/src/k8s.io/kubectl/pkg/cmd/drain/drain.go:
```diff
@@ -297,34 +298,45 @@ func (o *DrainCmdOptions) RunDrain() error {
-			break
+	deleteOrEvictWaitGroup.Wait()
+	remainingNodes := []string{}
+	for _, remainingInfo := range o.nodeInfos {
```
In light of the opt-in flag for parallel drain that I just pushed I think we still need this printing for the serial case. This will also print the list of nodes that failed to drain in the parallel case in a single line in the bottom. As you mentioned in your other comment the scrolling failure messages are easy to lose, so I think this final output is helpful in the parallel case too

soltysh replied:
I agree with Michael here, if this is not needed for parallel case, move this to the serial one, inside the `if-else` condition above.

mnussbaum replied:
The "are we parallel?" conditional above is actually being invoked in a loop over nodes, and this printing needs to happen only once. I can add another conditional on the same flag here though if you'd like?

I do think the absence of this output will have a negative impact on the UX in the parallel case but if you disagree I'm fine being overruled

mnussbaum commented:
I've pushed the changes for the parallel opt-in flag. Regarding the logging changes, would you like me to continue to iterate on that before seeing this merged?

If so, do you think adding the node name to all printed outputs invoked in the parallel portions would be best? Happy to continue to work on this.

michaelgugino reviewed (changes requested):
I think we should add some comments around the drain helper to indicate that it is now expected to be thread safe, otherwise we might regress on this feature in the future.

If parallel execution is opt-in, I don't think logging is necessarily a blocker for this PR.

mnussbaum committed:
Fix error aggregation during parallel node drain

A missing pass-by-reference was dropping errors from node drain

mnussbaum committed:
Add comments indicating functions that must be thread-safe

mnussbaum commented on staging/src/k8s.io/kubectl/pkg/cmd/drain/drain.go:
```diff
@@ -285,6 +287,28 @@ func (o *DrainCmdOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args [
+	drainResultMutex *sync.Mutex,
+	drainWG *sync.WaitGroup,
+	drainedNodes *sets.String,
+	drainErrors []error,
```
Ah thanks for spotting that. I've fixed this and tested to the error handling to confirm drain node errors cause kubectl to exit non-zero

mnussbaum commented:
@michaelgugino I've added comments about thread safety, let me know if you'd like more anywhere or updates to them

michaelgugino reviewed (commented):
This looks good to me overall.  I should have been more specific, but I think we should add the comment about thread-safety here: https://github.com/kubernetes/kubernetes/blob/491adb0dbc0e3832bc80922cbffbae4d4c72eb95/staging/src/k8s.io/kubectl/pkg/drain/drain.go#L48

That's the object that gets initialized and used by all the different threads in this patch, so if that object changes later to lose thread safety (eg, something mutates an internal field after initialization) then that might cause hard to detect race conditions in the future.

michaelgugino commented:
@soltysh minus a comment, I think this one is ready for your review.  I'm happy with the approach.

mnussbaum committed:
Fix thread-safety comment location

mnussbaum commented:
@michaelgugino thanks for the pointer, moved the thread-safety comment to the helper

mnussbaum commented:
/retest

michaelgugino reviewed (approved):
lgmt

mnussbaum commented:
/retest

mnussbaum commented:
/retest

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/drain/drain.go:
```diff
@@ -285,6 +287,28 @@ func (o *DrainCmdOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args [
+	drainedNodes *sets.String,
+	drainErrors *[]error,
+	printObj printers.ResourcePrinterFunc,
+) {
```
I'd prefer we split the parallelism part (workgroups & mutexes) outside this function, such that this function is only responsible for draining a node. The entire wrapper should be handled either inside a separate function or inside that `if-else` condition below.

mnussbaum replied:
The difficulty is that `drainNode` is being invoked in a go-routine, so the caller can't block waiting for it to return to update mutexes and WGs. I initially added this `drainNode` function to handle that parallelism book-keeping in a single place.

Would renaming `drainNode` make its book-keeping role more clear? Or do you think I should wrap `drainNode` in another function that will be invoked in the go-routine and will handle the book-keeping?

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/drain/drain.go:
```diff
@@ -194,6 +195,7 @@ func NewCmdDrain(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobr
 	cmd.Flags().StringVarP(&o.drainer.PodSelector, "pod-selector", "", o.drainer.PodSelector, "Label selector to filter pods on the node")
 	cmd.Flags().BoolVar(&o.drainer.DisableEviction, "disable-eviction", o.drainer.DisableEviction, "Force drain to use delete, even if eviction is supported. This will bypass checking PodDisruptionBudgets, use with caution.")
 	cmd.Flags().IntVar(&o.drainer.SkipWaitForDeleteTimeoutSeconds, "skip-wait-for-delete-timeout", o.drainer.SkipWaitForDeleteTimeoutSeconds, "If pod DeletionTimestamp older than N seconds, skip waiting for the pod.  Seconds must be greater than 0 to skip.")
+	cmd.Flags().BoolVar(&o.drainer.ParallelizeNodes, "parallelize-nodes", o.drainer.ParallelizeNodes, "Execute drains on multiple nodes in parallel")
```
I don't like this name that much, maybe just `--parallel`? Other option is to have something like `--num-workers` which would allow you to specify the number of parallel drains, wdyt?

mnussbaum replied:
Cool, I've updated the flag to be `--parallel`

soltysh reviewed (changes requested):
I left some more comments.

mnussbaum committed:
Rename --parallelize-nodes flag to --parallel
